### PR TITLE
Adds vmservices to retrieve android applink settings

### LINKED ISSUE #120408: [Deeplink Automation] Implements flutter tool to retrieve app settings from Android project
As describe in internal design doc [go/flutter-deeplink-web-validator](http://goto.google.com/flutter-deeplink-web-validator) (non-public accessible)

We need to build a tool to get package name, and associate web domain

### PR DESCRIPTION
**Pull request by chunhtai**
fixes https://github.com/flutter/flutter/issues/120408

Added two gradle tasks, one for grabing the application id, one for grabbing app link domains.

Added a new vmservices to call these two gradle tasks and return the result.

The expected work flow is that the devtool will first call a vmservices to grab all avaliable build variants. It will then choose one of the build variant and call this new services to get application id and app link domains.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by eliasyishak** [comment]
Hi, @chunhtai is this still a work in progress? Seems like you have some tests failing

**Comment by chunhtai** [comment]
These tests failures seem to be related to https://github.com/flutter/flutter/issues/126557

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy [hunk 1]
```diff
@@ -904,7 +976,11 @@ class FlutterPlugin implements Plugin<Project> {
             validateDeferredComponentsValue = project.property('validate-deferred-components').toBoolean()
         }
         addTaskForJavaVersion(project)
-        addTaskForPrintBuildVariants(project)
+        if(isFlutterAppProject()) {
```
[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/flutter_tools/gradle/src/main/groovy/flutter.groovy:982]: for add to app, these tasks can only be added in the main android project.

I think in the future I should refactor this out into a separate gradle plugin to support add-to-app, but I would like to make the decision later when we have more concrete design for add-to-app support in deeplink validation tool.

This don't need to be cherry pick as there isn't a client using these methods yet.

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy [hunk 2]
```diff
@@ -904,7 +979,11 @@ class FlutterPlugin implements Plugin<Project> {
...
+            addTaskForPrintBuildVariants(project)
```
[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/flutter_tools/gradle/src/main/groovy/flutter.groovy:983]: I think I still need the `project` to register task to?

[the remark this answers is not in the document]
Comment (reidbaker) [reply packages/flutter_tools/gradle/src/main/groovy/flutter.groovy:987]: ah if that is the case then ignore this comment.

File: packages/flutter_tools/lib/src/android/gradle.dart [hunk 3]
```diff
@@ -239,6 +272,38 @@ class AndroidGradleBuilder implements AndroidBuilder {
...
+    );
+    final List<String> command = <String>[
+      _gradleUtils.getExecutable(project),
+      ...options, // suppresses gradle output.
+      taskName,
+    ];
+
+    RunResult result;
+    try {
+      final String? javaHome = globals.androidSdk?.javaHome;
```
[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/flutter_tools/lib/src/android/gradle.dart]: I merged the changes

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy [hunk 4]
```diff
@@ -731,6 +731,85 @@ class FlutterPlugin implements Plugin<Project> {
...
+    //
+    // This task prints the application id in this format:
+    //
+    // ApplicationId: com.example.my_id
+    //
+    // Format of the output of this task is used by `AndroidProject.getApplicationIdForVariant`.
+    private static void addTasksForPrintApplicationId(Project project) {
+        project.android.applicationVariants.all { variant ->
+            // Warning: The name of this task is used by `AndroidProject.getApplicationIdForVariant`.
+            project.tasks.register("print${variant.name.capitalize()}ApplicationId") {
```
Comment (reidbaker) [inline packages/flutter_tools/gradle/src/main/groovy/flutter.groovy:745]: I think according to https://developer.android.com/reference/tools/gradle-api/7.4/com/android/build/api/variant/ApplicationVariant#applicationId() this depends on some task running that will merge the manifests. That means this task depends on another task (possibly as simple as assembleVariant). Given this I think you should add a dependsOn block as defined here. https://docs.gradle.org/current/userguide/tutorial_using_tasks.html#sec:task_dependencies

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy [hunk 5]
```diff
@@ -731,6 +731,81 @@ class FlutterPlugin implements Plugin<Project> {
...
+                description "Prints out application id for the given build variant of this Android project"
+                doLast {
+                    println "ApplicationId: ${variant.applicationId}";
+                }
+            }
+        }
+    }
+
+    // Add a task that can be called on Flutter projects that prints app link domains of a build
```
[the remark this answers is not in the document]
Comment (reidbaker) [reply packages/flutter_tools/gradle/src/main/groovy/flutter.groovy:754]: Great job on the documentation change here.

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy (hunk 4 above)
Comment (reidbaker) [reply packages/flutter_tools/gradle/src/main/groovy/flutter.groovy:745]: I think you might actually want `variant.outputs.each` see example here https://developer.android.com/build/gradle-tips#configure-dynamic-version-codes to print the applicationId. 
But I am not sure how task registration interacts differently if you use outputs.

Comment (reidbaker) [reply packages/flutter_tools/gradle/src/main/groovy/flutter.groovy:745]: FWIW I am trying to give good guidance here if you explore these things and they dont work you can ignore my comments.

**Review by reidbaker — APPROVED** [review]

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy (hunk 4 above)
Comment (chunhtai) [reply packages/flutter_tools/gradle/src/main/groovy/flutter.groovy:745]: It looks like it is not depending on any task, I ran it after a clean, and it still works. Also the application id is per applicationvariant I think? i didn't see a property on variantoutput for applicationID.

**Pull request merged**